File: fsharp/Bootstrap.Core/Pool.fs
namespace Bootstrap.Core

open System.Collections.Generic
open NghiaBui.Common

type 'T Pool when 'T : equality () =
    let seeds = HashSet<'T> ()
    let forbiddens = HashSet<'T> ()
    let iSeeds = IHS seeds
    let iForbiddens = IHS forbiddens

    member this.Seeds = iSeeds
    member this.Forbiddens = iForbiddens

    member this.AddSeeds (xs : seq<'T>) =
        seeds.UnionWith xs
        for x in xs do forbiddens.Remove x |> ignore
        this

    member this.AddForbiddens (xs : seq<'T>) =
        forbiddens.UnionWith xs
        for x in xs do seeds.Remove x |> ignore
        this

    member this.Remove (xs : seq<'T>) =
        for x in xs do
            seeds.Remove x |> ignore
            forbiddens.Remove x |> ignore
        this

    member this.Clear () =
        seeds.Clear ()
        forbiddens.Clear ()
        this

    member this.isNewAndValid x =
        not (seeds.Contains x) && not (forbiddens.Contains x)

File: java/src/com/nghiabui/bootstrap/core/Pool.java
package com.nghiabui.bootstrap.core;

import java.io.Serializable;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class Pool<T> implements Serializable {

	private final Set<T> seeds = new HashSet<>();
	private final Set<T> forbiddens = new HashSet<>();

	public Set<T> seeds() {
		return seeds;
	}

	public Set<T> forbiddens() {
		return forbiddens;
	}

	public boolean isSeed(T e) {
		return seeds.contains(e);
	}

	public boolean isForbidden(T e) {
		return forbiddens.contains(e);
	}

	public boolean isNewAndValid(T e) {
		return !isSeed(e) && !isForbidden(e);
	}

	public void addSeed(T e) {
		seeds.add(e);
		forbiddens.remove(e);
	}

	public void addSeeds(Collection<T> es) {
		seeds.addAll(es);
		forbiddens.removeAll(es);
	}

	public void addForbidden(T e) {
		seeds.remove(e);
		forbiddens.add(e);
	}

	public void addForbiddens(Collection<T> es) {
		seeds.removeAll(es);
		forbiddens.addAll(es);
	}

	public void remove(T e) {
		seeds.remove(e);
		forbiddens.remove(e);
	}

	public void remove(Collection<T> es) {
		seeds.removeAll(es);
		forbiddens.removeAll(es);
	}

	public void clear() {
		seeds.clear();
		forbiddens.clear();
	}

}
